### ISSUE TITLE
SkipLayerNormFusion -- High Output Difference Between PyTorch and ONNX Runtime with Extended Optimizations

### ISSUE THREAD

**Issue by WoodieDudy** [microsoft/onnxruntime#17689]
I have a trained PyTorch model, and when I export it  to onnx with `static shapes` in `fp16` and with default `all optimizations enabled` by ort.GraphOptimizationLevel.ORT_ENABLE_ALL.
When I compare the network outputs between PyTorch and ONNX Runtime, I observe a significant difference in the output tensors:
- Max diff: 9.7656
- Mean diff: 1.6991

### I get small difference:
1. When I set `optimization level to basic`

```python
sess_options = onnxruntime.SessionOptions()
sess_options.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_BASIC
ort_session = ort.InferenceSession(f"{model_name}.onnx", providers=providers, sess_options=sess_options)
```
2. When I export with `dynamic_shapes` and `GraphOptimizationLevel.ORT_ENABLE_All`
3. With just `fp32`

For cases above outputs diff looks like this:

- Max diff: 0.2656
- Mean diff: 0.0305

I think this is a bug in the extended optimization fuses of ORT or fp16<>fp32 casts.
This bug is only reproducible on a trained model, but I can't post it. But [here](https://drive.google.com/file/d/1GBqMt6sv497q64h88FoaqrlyAkrIrUzU/view?usp=sharing) example of random initialized model. Maybe it can help to understand precision casts inside

Thanks

### Platform

Linux

### OS Version

Ubuntu 22.04

### ONNX Runtime Version

onnxruntime-gpu==1.16.0

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

CUDA

### Execution Provider Library Version

CUDA Version: 12.1

**Comment by hariharans29** [comment]
1) So is the diff is low (side question - is low == acceptable or just lower diff than the "bad" case?) when using either fp32 or exporting with dynamic shapes ?

2) I am guessing there is a bug with fp16 extended optimization and static shapes probably induces an optimizer into action which would otherwise not be in play because the shape is not known statically. 

This might take some time to track down. One quick way to continue investtigating is try disabling optimizers in the EXTENDED suite (see comment here - https://github.com/microsoft/onnxruntime/issues/17476#issuecomment-1712967164) and narrow it down to an optimizer that is causing the problem.

**Comment by vadimkantorov** [comment]
@hariharans29 Indeed, for this kind of faulty-optimizer debugging it would be useful to get a list of enabled optimizer names per a given `opt_level` / `graph_optimization_level`. Otherwise it requires some code digging to feed `disabled_optimizers` name list...

**Comment by WoodieDudy** [comment]
1. low diff is acceptable
2. Can you please provide a list of optimizers that need to be disabled

**Comment by hariharans29** [comment]
@vadimkantorov  /  @WoodieDudy - 

1) ....it would be useful to get a list of enabled optimizer names per a given opt_level / graph_optimization_level

Same ask here - https://github.com/microsoft/onnxruntime/issues/17476. No API is available currently to retrieve the list of optimizers. We will need to prirotize the request.

2) Can this disabled_optimizers be fed directly into the session options?

See example usage [here](https://github.com/microsoft/onnxruntime/blob/1f4a3529ddde7796cd018b6c2bd97a6e099c4b88/onnxruntime/python/tools/quantization/quant_utils.py#L624)

3) Can you please provide a list of optimizers that need to be disabled
Extended optimizers or Level 2 optimizers (which is where you are seeing the bug manifest)  should be listed here - https://github.com/microsoft/onnxruntime/blob/1f4a3529ddde7796cd018b6c2bd97a6e099c4b88/onnxruntime/core/optimizer/graph_transformer_utils.cc#L285. Based on usage example above, to disable fusion of Gemm + Activation use "GemmActivationFusion". There seems to be [a check](https://github.com/microsoft/onnxruntime/blob/1f4a3529ddde7796cd018b6c2bd97a6e099c4b88/onnxruntime/core/session/inference_session.cc#L785) to ensure that the optimizers that are being disabled are available in the build - . Keep in mind that some Level 2 optimizers may not be applicable to the CUDA EP. For example, [this](https://github.com/microsoft/onnxruntime/blob/1f4a3529ddde7796cd018b6c2bd97a6e099c4b88/onnxruntime/core/optimizer/graph_transformer_utils.cc#L290) is only compatible with the CPU EP. 

I understand this is not the most ideal way to debug this but is the best way I can think of to get started with the debugging.

**Comment by wschin** [comment]
I feel there should be a flag such as `DUMP_GRAPH_BEFORE_AND_AFTER_TRANSFORM` to automatically get the graph before and after each applied transforms.

**Comment by WoodieDudy** [comment]
@hariharans29
I took a list of optimizers from  [here](https://github.com/microsoft/onnxruntime/blob/1f4a3529ddde7796cd018b6c2bd97a6e099c4b88/onnxruntime/core/optimizer/graph_transformer_utils.cc#L285), ran several experiments to find out which optimizers turn off reduces the error

It turned out that disabling `SkipLayerNormFusion` reduces the error by such values:
max difference:  9.7656 -> 0.3906
mean difference:  1.6991 -> 0.0758


```python
sess_options = onnxruntime.SessionOptions()
# sess_options.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_ENABLE_BASIC
sess_options.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_ENABLE_ALL
self.onnxruntime_session = onnxruntime.InferenceSession(
    onnx_model_path,
    providers=providers,
    sess_options=sess_options,
    disabled_optimizers=[
        # 'QDQS8ToU8Transformer',
        # 'QDQSelectorActionTransformer',
        # 'ConvActivationFusion',
        # 'GeluFusion',
        # 'LayerNormFusion',
        # 'SimplifiedLayerNormFusion',
        # 'AttentionFusion',
        # 'EmbedLayerNormFusion',
        # 'GatherToSplitFusion',
        # 'GatherToSliceFusion',
        # 'MatmulTransposeFusion',
        # 'BiasGeluFusion',
        # 'FastGeluFusion',
        # 'GeluApproximation'
        'SkipLayerNormFusion',
        # 'QuickGeluFusion',
    ]
)
```

**Comment by hariharans29** [comment]
> @hariharans29 I took a list of optimizers from [here](https://github.com/microsoft/onnxruntime/blob/1f4a3529ddde7796cd018b6c2bd97a6e099c4b88/onnxruntime/core/optimizer/graph_transformer_utils.cc#L285), ran several experiments to find out which optimizers turn off reduces the error
> 
> It turned out that disabling `SkipLayerNormFusion` reduces the error by such values: max difference: 9.7656 -> 0.3906 mean difference: 1.6991 -> 0.0758
> 
> ```python
> sess_options = onnxruntime.SessionOptions()
> # sess_options.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_ENABLE_BASIC
> sess_options.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_ENABLE_ALL
> self.onnxruntime_session = onnxruntime.InferenceSession(
>     onnx_model_path,
>     providers=providers,
>     sess_options=sess_options,
>     disabled_optimizers=[
>         # 'QDQS8ToU8Transformer',
>         # 'QDQSelectorActionTransformer',
>         # 'ConvActivationFusion',
>         # 'GeluFusion',
>         # 'LayerNormFusion',
>         # 'SimplifiedLayerNormFusion',
>         # 'AttentionFusion',
>         # 'EmbedLayerNormFusion',
>         # 'GatherToSplitFusion',
>         # 'GatherToSliceFusion',
>         # 'MatmulTransposeFusion',
>         # 'BiasGeluFusion',
>         # 'FastGeluFusion',
>         # 'GeluApproximation'
>         'SkipLayerNormFusion',
>         # 'QuickGeluFusion',
>     ]
> )
> ```

Thanks for the investigation thus far. I guess we have some issue in the SLN fusion and/or implementation. CC: @tianleiwu.

**Comment by wschin** [comment]
@hariharans29, anyone you'd recommend to fix this? Please feel free to assign this issue. This looks like an important operator.

**Comment by hariharans29** [comment]
@wschin - I am unable to recommend anyone to fix this. Maybe @tianleiwu can comment.

Also from OP's comment - This bug is only reproducible on a trained model, but I can't post it. But [here](https://drive.google.com/file/d/1GBqMt6sv497q64h88FoaqrlyAkrIrUzU/view?usp=sharing) example of random initialized model. Maybe it can help to understand precision casts inside

Not sure if the posted model repros the SLN issue as the original issue was posted with the following comment (`I think this is a bug in the extended optimization fuses of ORT or fp16<>fp32 casts.`). @WoodieDudy can confirm if the posted model repros the issue. It would be easier for the investigator if the model can be trimmed down to a minimal repro than such a large model.

**Comment by tianleiwu** [comment]
@WoodieDudy, could you try older version like 1.15.1? If older version is good, it is a regression in 1.16 (we had two commits of SLN in 1.16).

**Comment by tianleiwu** [comment]
I looked at bugmodel.onnx, the SkipLayerNormalization only involves 4 inputs (input, skip, weight and bias). input shape is 16x4x512. There is no data Cast happens. The pattern is normal. 

I could do some test with random inputs to see whether SkipLayerNormalization kernel has precision issue.

**Comment by tianleiwu** [comment]
@WoodieDudy, could you try add a flag {"enable_skip_layer_norm_strict_mode": True} to provider option. Example usage: https://github.com/microsoft/onnxruntime/blob/a441a71e8e1b9f8cd1fa83cbc4d6aa5f4bec8e2f/onnxruntime/python/tools/transformers/convert_generation.py#L671-L677

Let me know whether that could reduce output difference.

**Comment by WoodieDudy** [comment]
@tianleiwu @hariharans29 
Sorry it's taken me a while to respond
Here's [code](https://github.com/WoodieDudy/nemo-conformer-inference-benchmarks/tree/main/ort_optimizer_bug) for repro.
There's also a link to weights on which the bug is reproduced

btw {"enable_skip_layer_norm_strict_mode": True}  solves the problem too

**Comment by tianleiwu** [comment]
It is likely caused by SkipLayerNorm kernel using fp16 in some part of accumulation.
{"enable_skip_layer_norm_strict_mode": True}  will force to use fp32 in accumulation, but that will get worse performance.
There is trade off between performance and accuracy there.

**Comment by vadimkantorov** [comment]
@tianleiwu is there any weight recomputation happening during fusion? Is it by chance done in fp16?

Otherwise, I see your idea...

**Comment by tianleiwu** [comment]
@vadimkantorov, there is NO weight recomputation during fusion. See the source code:
https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/optimizer/skip_layer_norm_fusion.cc

The acc dtype is hard coded in cuda kernel and it is not related to fusion.

**Comment by vadimkantorov** [comment]
Does it mean that acc dtype is actually different in regular LayerNorm and if SkipLayerNormFusion is enabled? (model weights and inputs are fp16 for both cases)
